Document advanced query capabilities on Azure AD objects (#12832)
microsoftgraph/microsoft-graph-docs: concepts/aad-advanced-queries.md

---
title: "Advanced query capabilities on Azure AD directory objects"
description: "Azure AD directory objects support advanced queries based on the OData query parameters."
author: "FaithOmbongi"
localization_priority: Priority
ms.custom: graphiamtop20, scenarios:getting-started
---

# Advanced query capabilities on Azure AD directory objects

> As Azure AD continues to deliver more capabilities and improvements in stability, availability, and performance, Microsoft Graph also continues to evolve and scale to efficiently access the data. One way is through Microsoft Graph's increasing support for advanced query capabilities on various Azure AD objects and their properties. For example, the addition of **Not** (`NOT`), **Not equals** (`ne`), and **Ends with** (`endsWith`) operators on the `$filter` query parameter in October 2020.

Reviewer: "in October 2020" isn't needed.

Author: Was thinking of this in the context of "improvements, evolution, and scalability" pointed out in the 1st sentence.

| Use of `$filter` with `NOT` operator                                                                                                                           | [GET](https://developer.microsoft.com/graph/graph-explorer?request=users%3F%24filter%3DNOT%20startsWith(displayName%2C%20'Luca')%26%24count%3Dtrue&method=GET&version=beta&GraphUrl=https://graph.microsoft.com&headers=W3sibmFtZSI6IkNvbnNpc3RlbmN5TGV2ZWwiLCJ2YWx1ZSI6ImV2ZW50dWFsIn1d) `../users?$filter=NOT startsWith(displayName, 'Luca')&$count=true`                                                                         |

> [!NOTE]
> These advanced query capabilities are not available in Azure AD B2C tenants.

## Support for filter on properties of Azure AD directory objects

> Properties of directory objects behave differently in their support for query parameters. The following table summarizes support for `$filter` operators by properties on the [users](/graph/api/resources/user) directory object. Properties for other directory objects will be added soon.

Reviewer: We shouldn't include text that implies promises of future functionality: "Properties for other directory objects will be added soon."

> - ![Requires Advanced Query Parameters.](/graph/images/advanced-query-parameters/AQP.png) The property supports `$filter` with the operator by default.

Reviewer: Since advanced query parameters isn't a brand name, I don't think the term should be capitalized.

- ![Works by default. Does not require Advanced Query Parameters.](/graph/images/advanced-query-parameters/default.png) The specific `$filter` operator requires *Advanced Query Parameters*:
  - `ConsistencyLevel=eventual` header
  - `$count=true` query string
- Blank cells indicate that the property does not support the use of `$filter` with the operator.
- The null value column indicates that the property is filterable on `null` values.
- Properties that are not listed here do not support `$filter`.

> > [!NOTE]

Reviewer: I don't think the note format is necessary. And the bullet items are not steps to complete so shouldn't be numbered.

Author: Well noted. The patterns have also grown from the origin 2.

>
> 1. Properties with the same name across directory resources support the same `$filter` operators. For example, the **createdDateTime** property is available in **application**, **group**, **organization**, and **user** resources. It supports the `eq`, `ge`, and `le` operators by default and the `in` and `ne` operators only in advanced queries.
> 2. `endsWith` operator is supported only on **mail** and **userPrincipalName** properties.
> 3. Queries that are supported by default will also work in advanced queries.
> 3. `NOT` and `ne` operators are supported only in advanced queries.
>>

[!INCLUDE [filter-directory-objects](../includes/filter-directory-objects.md)]

## Error handling for advanced queries on directory objects

> Counting Directory Objects is only supported using the advanced queries parameters.  

Reviewer: ```suggestion
Counting directory objects is only supported using the advanced queries parameters.  
```

> If the `ConsistencyLevel=eventual` header is not specified, the request will return an error in case raw count is used (`/$count` query segment) or silently ignoring the count parameter in case inline count is used (`$count=true` query parameter).

Reviewer: ```suggestion
If the `ConsistencyLevel=eventual` header is not specified, the request returns an error when the $count segment is used or silently ignores the $count query parameter (?$count=true) if it's used.
```

Author: Added change in local branch.

The Microsoft Graph query engine uses an index store to fulfill query requests. To add support for additional query capabilities on some properties, these properties are now indexed in a separate server. This separate indexing allows Azure AD to increase support and improve the performance of the query requests. However, these advanced query capabilities are not available by default but, the requestor must also set the **ConsistencyLevel** header with the value `eventual` *and*, with the exception of `$search`, use the `$count` query parameter (either as a [URL segment](/graph/query-parameters#other-odata-url-capabilities) or `$count=true` query string). The **ConsistencyLevel** header and `$count` are referred to as *advanced query parameters*.

For example, if you wish to retrieve only inactive user accounts, you can run either of these queries that use the `$filter` query parameter.

> 1. Use the `$filter` query parameter with the `eq` operator. This request will work by default, that is, the request does not require the advanced query parameters.

Reviewer: These should not be numbered. Numbering implies a sequential order of operation.

The following table summarizes the `$filter` operators support, as properties of directory objects behave differently.

- ⚪ indicates that the property supports `$filter` with the operator by default.
- ⚫ indicates that the specific `$filter` operator requires the *Advanced Queries Parameters*:
  - `ConsistencyLevel=eventual` header
  - `$count=true` query string
- Blank cells indicate that the property does not support the use of `$filter` with the operator.
- Null value column indicates that the property is filterable using `null`
- Properties that are not listed here do not support `$filter`.

> [!NOTE]
>
> 1. Properties with the same name across directory resources support the same `$filter` operators. For example, the **createdDateTime** property is available in **application**, **group**, **organization**, and **user** resources. It supports the `eq`, `ge`, and `le` operators by default and the `in` and `startsWith` operators only in advanced queries.
> 2. Queries that are supported by default will also work in advanced queries.
> 3. `NOT` and `ne` operators are only supported in advanced queries.
>>

> | Property                      |  eq   |  ne   |  ge   |  le   |  in   | startsWith | endsWith | null value |

Reviewer 2: @FaithOmbongi null value indexing is only supported with Advanced Queries.
for example, [this query](https://developer.microsoft.com/en-us/graph/graph-explorer?request=users%3F%24count%3Dtrue%26%24filter%3Dcountry%20eq%20null%26%24select%3Did%2CdisplayName%2Ccountry&method=GET&version=beta&GraphUrl=https://graph.microsoft.com) doesn't work without mezzo parameters.

We also need to find a way to indicate whenever the "null default value" is implemented (e.g. for accountEnabled all nulls fall back to 'false' value). Null indexing and Null default value are mutually exclusive and they can be represented in the same column.

| :---------------------------- | :---: | :---: | :---: | :---: | :---: | :--------: | :------: | :--------: |

> | accountEnabled                |   ⚪   |   ⚫   |   ⚫   |   ⚫   |   ⚪   |            |          |            |

Reviewer 2: using only colored dots is not a best practice for accessibility, can we find symbols that have different shapes? (also considering the 3rd one regarding null indexes / null default values.

Author: Let me revisit the accessibility guidelines. I see [Azure DB docs use ticks](https://azure.microsoft.com/en-us/product-categories/databases/).

- ⚪ The property supports `$filter` with the operator by default.
- ⚫ The specific `$filter` operator requires the *Advanced Query Parameters*:
  - `ConsistencyLevel=eventual` header
  - `$count=true` query string
- Blank cells indicate that the property does not support the use of `$filter` with the operator.
- The null value column indicates that the property is filterable on `null` values.
- Properties that are not listed here do not support `$filter`.

> [!NOTE]
>
> 1. Properties with the same name across directory resources support the same `$filter` operators. For example, the **createdDateTime** property is available in **application**, **group**, **organization**, and **user** resources. It supports the `eq`, `ge`, and `le` operators by default and the `in` and `ne` operators only in advanced queries.
> 2. Queries that are supported by default will also work in advanced queries.
> 3. `NOT` and `ne` operators are supported only in advanced queries.
>>

> | Property                      | eq | ne | NOT | ge | le | in | startsWith | endsWith | null values |

Reviewer 2: ne and NOT are **always** supported with advanced queries, probably not useful to call out in the table since we told in the line above?

Author: Anything to simplify the table and group shared scenarios would be good.

1. But, do all properties support NOT and ne?
2. Or, can we say a common rule is that all properties that support eq also support NOT and ne? With a note that ne cannot be used inside any/all clauses (hit this error when I tried)

> | Property                      | eq | ne | NOT | ge | le | in | startsWith | endsWith | null values |

Reviewer 2: we need to find a way to clarify when `$filter=property eq null` is supported
and when `$filter=property eq false` actually means false or null.

Author: We can discuss this further.

| postalCode                    | ⚫  | ⚫  |     | ⚫  | ⚫  | ⚫  | ⚫          |          | ⚫           |
| preferredLanguage             | ⚫  | ⚫  |     | ⚫  | ⚫  | ⚫  | ⚫          |          | ⚫           |
| proxyAddresses                | ⚪  | ⚫  |     | ⚪  | ⚪  |    | ⚪          |          | ⚫           |
| showInAddressList             | ⚫  | ⚫  |     | ⚫  | ⚫  | ⚫  |            |          |             |
| signInActivity                | ⚪  | ⚫  |     | ⚪  | ⚪  |    |            |          |             |
| state                         | ⚪  | ⚫  |     | ⚪  | ⚪  | ⚪  | ⚪          |          | ⚫           |
| streetAddress                 | ⚫  | ⚫  |     | ⚫  | ⚫  | ⚫  | ⚫          |          | ⚫           |
| surname                       | ⚪  | ⚫  |     | ⚪  | ⚪  | ⚪  | ⚪          |          | ⚫           |
| usageLocation                 | ⚪  | ⚫  |     | ⚪  | ⚪  | ⚪  | ⚪          |          |             |
| userPrincipalName             | ⚪  | ⚫  |     |    | ⚪  | ⚪  | ⚪          | ⚫        |             |
| userType                      | ⚪  |    |     |    |    | ⚪  |            | ⚫        | ⚫           |

> 

Reviewer 2: we should add a similar table for the orderBy support

Author: I think we can add that to the individual property descriptions.
